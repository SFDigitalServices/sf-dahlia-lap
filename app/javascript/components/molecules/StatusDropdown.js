import React from 'react'

import { LEASE_UP_STATUS_OPTIONS, getLeaseUpStatusClass } from '~/utils/statusUtils'
import Dropdown from '../molecules/Dropdown'

const StatusDropdown = ({ status, onChange, buttonClasses, menuClasses, wrapperClasses, styles, disabled }) => {
  let dropdownButtonClasses = [getLeaseUpStatusClass(status)]
  if (buttonClasses) dropdownButtonClasses.push(...buttonClasses)

  return (
    <Dropdown
      items={LEASE_UP_STATUS_OPTIONS}
      value={status}
      prompt='Status'
      onChange={onChange}
      buttonClasses={dropdownButtonClasses}
      menuClasses={menuClasses}
      wrapperClasses={wrapperClasses}
      styles={styles}
      disabled={disabled} />
  )
}

export default StatusDropdown
