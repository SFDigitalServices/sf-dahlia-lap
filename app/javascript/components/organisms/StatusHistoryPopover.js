import React, { useState } from 'react'

import classNames from 'classnames'
import PropTypes from 'prop-types'

import apiService from 'apiService'
import { COLORS } from 'components/atoms/colors'
import StyledIcon from 'components/atoms/StyledIcon'
import StatusItems from 'components/molecules/lease_up_sidebar/StatusItems'
import Popover from 'components/molecules/Popover'

import Loading from '../molecules/Loading'

const MAX_UPDATES_TO_SHOW_DEFAULT = 4

const StatusHistoryPopover = ({ applicationId, customButtonClasses = null }) => {
  const [showingAllStatuses, setShowingAllStatuses] = useState(false)
  const [state, setState] = useState({ statusItems: [], loading: false })

  const onShowHideStatusesToggled = () => setShowingAllStatuses(!showingAllStatuses)

  const fetchStatusHistory = () => {
    setState({ statusItems: state.statusItems, loading: true })
    apiService
      .getFieldUpdateComments(applicationId)
      .then((items) => setState({ statusItems: items, loading: false }))
      .catch(() => setState({ statusItems: state.statusItems, loading: false }))
  }

  const onClick = () => {
    state.statusItems.length === 0 && fetchStatusHistory()
  }

  const numberOfStatusesToDisplay = showingAllStatuses
    ? state.statusItems.length
    : MAX_UPDATES_TO_SHOW_DEFAULT

  const showShowMoreToggler = state.statusItems.length > MAX_UPDATES_TO_SHOW_DEFAULT
  return (
    <Popover
      buttonElement={({ onClick, ref }) => (
        <button
          ref={ref}
          onClick={onClick}
          className={classNames(customButtonClasses, 'button-icon-only')}
        >
          <StyledIcon icon='list-unordered' customSizeRem='1.2rem' customFill={COLORS.steel} />
        </button>
      )}
      onButtonClick={onClick}
    >
      <Loading isLoading={state.loading}>
        <StatusItems
          statusItems={state.statusItems}
          limit={numberOfStatusesToDisplay}
          height={showShowMoreToggler ? '20rem' : null}
        />
        {showShowMoreToggler && (
          <button
            className={classNames('button-link', 't-tiny')}
            type='button'
            onClick={onShowHideStatusesToggled}
          >
            {showingAllStatuses ? 'Show only recent status updates' : 'Show all status updates'}
          </button>
        )}
      </Loading>
    </Popover>
  )
}
StatusHistoryPopover.propTypes = {
  applicationId: PropTypes.string.isRequired,
  customButtonClasses: PropTypes.string
}

export default StatusHistoryPopover
