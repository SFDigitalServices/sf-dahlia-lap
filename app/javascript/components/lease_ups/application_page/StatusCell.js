import React from 'react'

import { PropTypes } from 'prop-types'

import StatusDropdown from 'components/molecules/StatusDropdown'
import StatusHistoryPopover from 'components/organisms/StatusHistoryPopover'
import { LEASE_UP_STATUS_VALUES } from 'utils/statusUtils'

export const VALIDATION_CONFIRMED = 'Confirmed'
export const VALIDATION_UNCONFIRMED = 'Unconfirmed'
export const VALIDATION_INVALID = 'Invalid'

const cellContainerStyle = {
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'flex-end',
  width: '100%'
}

const StatusCell = ({ applicationId, status = null, onChange = (_) => {} }) => {
  return (
    <div style={cellContainerStyle}>
      <StatusDropdown status={status} size='tiny' onChange={onChange} />
      <StatusHistoryPopover applicationId={applicationId} customButtonClasses='margin-left' />
    </div>
  )
}

StatusCell.propTypes = {
  applicationId: PropTypes.string.isRequired,
  status: PropTypes.oneOf(LEASE_UP_STATUS_VALUES),
  onChange: PropTypes.func
}

export default StatusCell
