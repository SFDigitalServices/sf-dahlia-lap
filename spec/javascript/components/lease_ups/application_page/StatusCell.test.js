import React from 'react'

import { shallow } from 'enzyme'

import StatusCell from 'components/lease_ups/application_page/StatusCell'
import StatusDropdown from 'components/molecules/StatusDropdown'
import StatusHistoryPopover from 'components/organisms/StatusHistoryPopover'

const mockOnChange = jest.fn()

describe('StatusCell', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(
      <StatusCell applicationId='applicationId1' status='Waitlisted' onChange={mockOnChange} />
    )
  })

  test('renders an a dropdown and a popover', () => {
    expect(wrapper.find(StatusDropdown)).toHaveLength(1)
    expect(wrapper.find(StatusHistoryPopover)).toHaveLength(1)
  })

  test('renders the popover button with left margin', () => {
    expect(wrapper.find(StatusHistoryPopover).props().customButtonClasses).toEqual('margin-left')
  })

  test('onClick is trigged when the input changes', () => {
    expect(mockOnChange.mock.calls).toHaveLength(0)
    wrapper.find(StatusDropdown).simulate('change')
    expect(mockOnChange.mock.calls).toHaveLength(1)
  })
})
