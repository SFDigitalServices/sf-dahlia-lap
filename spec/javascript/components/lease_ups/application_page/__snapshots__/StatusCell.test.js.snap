// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusCell renders with the correct styling 1`] = `
<DocumentFragment>
  <div
    style="display: flex; align-items: center; justify-content: flex-end; width: 100%;"
  >
    <div
      class="dropdown css-b62m3t-container"
    >
      <span
        class="css-1f43avz-a11yText-A11yText"
        id="react-select-3-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        class="css-1f43avz-a11yText-A11yText"
        role="log"
      />
      <div
        class="status-dropdown__control css-uuqjpg-Control"
      >
        <button
          class="button dropdown-button is-waitlisted tight-padding tiny margin-bottom-none"
          style="padding-top: 0px; padding-bottom: 0px;"
          type="button"
        >
          <div
            style="display: flex; align-items: center; justify-content: space-between; width: 100%;"
          >
            <span
              style="margin-top: 0.5rem; margin-bottom: 0.5rem; margin-right: 0.5rem; overflow: auto; flex-grow: 1; text-align: left;"
            >
              Waitlisted
            </span>
            <div
              style="width: 0px; height: 0px;"
            >
              <div
                class="status-dropdown__value-container status-dropdown__value-container--has-value css-1fdsijx-ValueContainer"
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-readonly="true"
                  class="css-mohuvp-dummyInput-DummyInput"
                  id="react-select-3-input"
                  inputmode="none"
                  role="combobox"
                  tabindex="0"
                  value=""
                />
              </div>
            </div>
            <span
              style="flex-shrink: 0; font-size: 0px;"
            >
              <svg
                style="width: .75rem; height: .75rem;"
              >
                <use
                  xlink:href="#i-arrow-down"
                />
              </svg>
            </span>
          </div>
        </button>
      </div>
    </div>
    <button
      class="margin-left button-icon-only"
    >
      <svg
        style="width: 1.2rem; height: 1.2rem; fill: #767676;"
      >
        <use
          style="fill: #767676;"
          xlink:href="#i-list-unordered"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
